File: Medium/322.Coin_Change.java
package Medium;

import java.util.Arrays;

class Solution {
    public int coinChange(int[] coins, int amount) {
        if (amount == 0) {
            return 0;
        }
        
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, Integer.MAX_VALUE);        
        dp[0] = 0;

        for (int i = 0; i < amount + 1; i++) {
            if (dp[i] == Integer.MAX_VALUE) {
                continue;
            }

            for (int coin : coins) {
                if (0 < i + coin && i + coin < amount + 1) {
                    dp[i + coin] = Math.min(dp[i + coin], dp[i] + 1);
                }
            }
        }

        return dp[amount] == Integer.MAX_VALUE ? -1 : dp[amount];
    }
}

File: Python/Medium/322.CoinChange.py
from collections import defaultdict
from typing import NewType

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        self.c = defaultdict(lambda: -1)
        count = self.recursion(coins, amount)
        if count == 9999:
            return -1
        return count

    def recursion(self, coins, amount):
        if amount == 0:
            return 0
        elif amount > 0:
            if self.c[amount] == -1:
                count = 9999
                for n in coins:
                    count = min(count, self.recursion(coins, amount - n) + 1)
                self.c[amount] = count
                return count
            else:
                return self.c[amount]
        else:
            return 9999    
